src/main/java/com/google/sps/servlets/AddSchoolInfo.java

```diff
@@ -0,0 +1,159 @@
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import java.io.IOException;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/add-info")
+public class AddSchoolInfo extends HttpServlet {
+
+  private Key existingSchoolKey;
+  private Key existingCourseKey;
+  private Key existingProfessorKey;
+
+  private Key newSchoolKey;
+  private Key newCourseKey;
+  private Key newProfessorKey;
+  private Key newTermKey;
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    String schoolName = request.getParameter("school-name");
+    String courseName = request.getParameter("course-name");
+    String term = request.getParameter("term");
+    Integer units = Integer.parseInt(request.getParameter("units"));
+    String profName = request.getParameter("professor-name");
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    Filter schoolFilter = new FilterPredicate("school-name", FilterOperator.EQUAL, schoolName);
+    Query schoolQ = new Query("school").setFilter(schoolFilter);
+    List<Entity> schoolQuery =
+        datastore.prepare(schoolQ).asList(FetchOptions.Builder.withDefaults());
+
+    if (schoolQuery.isEmpty()) {
+      Entity newSchool = new Entity("school");
+      newSchool.setProperty("school-name", schoolName);
+      newSchoolKey = newSchool.getKey();
+      datastore.put(newSchool);
+    } else {
+      Entity existingSchool = schoolQuery.get(0);
+      existingSchoolKey = existingSchool.getKey();
+      newSchoolKey = null;
+    }
+
+    if (newSchoolKey != null) {
+      Entity newCourse = new Entity("course", newSchoolKey);
+      newCourse.setProperty("course-name", courseName);
+      newCourse.setProperty("units", units);
+      newCourseKey = newCourse.getKey();
+      datastore.put(newCourse);
+
+      Entity newProfessor = new Entity("professor", newSchoolKey);
+      newProfessor.setProperty("professor-name", profName);
+      newProfessorKey = newProfessor.getKey();
+      datastore.put(newProfessor);
+
+      Entity newTerm = new Entity("term", newCourseKey);
+      newTerm.setProperty("term", term);
+      newTerm.setProperty("professor", newProfessorKey);
+      datastore.put(newTerm);
```

Comment:
You're handling everything at once it seems (course, professor, and term).
While this works, it means there's a lot of code to traverse to get to the details of how the other term creations for example are used and it's easy to miss an edge case.

I'd advise splitting this up into three functions: `getOrCreateCourse`, `getOrCreateProfessor`, and `getOrCreateTerm`.
Then create some booleans if needed like `isNewSchool`, `isNewProfessor`, `isNewTerm` etc. With corresponding functions as needed.

Have the functions pass back the Entity created, then use that for your final data store puts at the end.

Having a conceptual:
- Create course
- Create professor
- Create term
- Store them

With necessary logic aggregated as you go via booleans/etc I think would be clearer.

It's also possible I'm misunderstanding this and you need this extra complexity/branching. If so ping me and we can chat about it further.